### PR TITLE
testing mock in phpspec

### PR DESCRIPTION
**Pull request by ghamoron**
Mock phpspec

### CODE REVIEW

**Comment by samsonasik** [comment]
it conflict with master branch, I updated the spec file in `spec/App` directory, please do rebase against upstream master. Check https://samsonasik.wordpress.com/2015/09/16/practical-git-4-rebasing-conflicted-task-branch-against-primary-branch/ for how to do rebase ;)

**Comment by samsonasik** [comment]
at this case, you can do:

``` shell
$ git checkout master
$ git remote add upstream https://github.com/sitrunlab/learn-phpspec.git
$ git pull upstream master
$ git checkout first_learn_phpspec
$ git rebase master
```

You will get conflict, fix the "<<<<<<< HEAD" mark, move the spec file to under `spec/App` directory and `src/*.php` to `src/App/` like I described above, and remove original file under `spec` directory, the namespace of spec will be:

``` php
namespace spec\App;
```

with lower `spec` instead of : 

``` php
namespace Spec\App;
```

After that, you can run:

``` shell
$ git add .
$ git rebase --continue
$ git push --force origin first_learn_phpspec
```

and your PR will be synchronized. Please let me know if you need help for it ;)

**Comment by samsonasik** [comment]
after you rebase, please run `composer dump-autoload` as the psr4 autoload now uses dedicated `App` folder under `src`.

**Comment by ghamoron** [comment]
Try rebase without any conflict or error. Thank's :)

**Comment by samsonasik** [comment]
cool, now we can see travis error as expected, just add more commits until it green ;).

File: spec/App/StubMarkdownSpec.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+* @param App\StubReader $reader
+*/
+
+class StubMarkdownSpec extends ObjectBehavior
+{
+    function it_converts_text_from_an_external_source(StubReader $reader)
+    {
+    	$reader->getMarkdown()->willReturn("<p>Hi, there!</p>");
+
+    	$this->toHtmlFromReader($reader)->shouldReturn("<p>Hi, there!</p>");
```
Comment (samsonasik) [inline spec/App/StubMarkdownSpec.php]: just move the method to `MarkdownSpec` and remove this class

File: src/App/StubMarkdown.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+namespace App;
+
+class StubMarkdown
+{
+    public function toHtmlFromReader($argument1)
+    {
+        return $argument1->getMarkdown();
+    }
```
Comment (samsonasik) [inline src/App/StubMarkdown.php]: move this method to Markdown class, and remove this class

File: src/App/StubReader.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace App;
+
+interface StubReader
```
Comment (samsonasik) [inline src/App/StubReader.php]: rename it to `Reader`, or `ReaderInterface`. Stub is a process in test that record info about function call.

File: src/App/Markdown.php [hunk 4]
```diff
@@ -21,4 +21,8 @@
 
 class Markdown
 {
+    public function outputHtml($argument1, $argument2)
```
Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` typehint to 2nd argument

File: src/App/InvalidArgumentException.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace App;
+
+class InvalidArgumentException
```
Comment (samsonasik) [inline src/App/InvalidArgumentException.php]: add extends: `class InvalidArgumentException extends \InvalidArgumentException` to make it an exception class. 
I prefer to use  add use statement first and make alias.

``` php
use InvalidArgumentException as BaseInvalidArgumentException;

class InvalidArgumentException extends BaseInvalidArgumentException
```

File: src/App/Markdown.php [hunk 6]
```diff
@@ -21,4 +21,29 @@
 
 class Markdown
 {
+	// Mock test
```
Comment (samsonasik) [inline src/App/Markdown.php]: remove this comment, "Mock" happen in test

File: src/App/Markdown.php [hunk 7]
```diff
@@ -21,4 +21,29 @@
...
+
+        return $markdown;
+    }
+
+    public function outputHtml($argument1, $argument2)
+    {
+        $argument2->writeText($argument1);
+    }
+
+    // Stub test
```
Comment (samsonasik) [inline src/App/Markdown.php]: remove this comment, "Stub" happen in test

File: spec/App/MarkdownSpec.php [hunk 8]
```diff
@@ -21,11 +21,34 @@
 
 use PhpSpec\ObjectBehavior;
 use Prophecy\Argument;
+use App\Writer;
+use App\Reader;
 
 class MarkdownSpec extends ObjectBehavior
 {
-    function it_is_initializable()
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php:27]: `it_is_initializable()` with `$this->shouldHaveType('App\Markdown');` content still ok here, no need to remove the method.

File: src/App/Markdown.php [hunk 9]
```diff
@@ -21,4 +21,29 @@
...
 {
+	// Mock test
+    public function __construct($argument1)
+    {
+        // TODO: write logic here
+    }
+
+    public static function createForWriting($argument1)
+    {
+        $markdown = new Markdown($argument1);
```
Comment (samsonasik) [inline src/App/Markdown.php]: `$markdown = new static($argument1);` as no indication the class won't be extended.

File: src/App/Movie.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+        return $desc;
+    }
+
+    public function setRating($argument1)
+    {
+        if($argument1 < 0)
+        {
+        	throw new InvalidArgumentException("Error Processing Request");	
+        }
```
Comment (samsonasik) [inline src/App/Movie.php:62]: use psr-2 coding style guide, see https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#51-if-elseif-else

File: spec/App/MarkdownSpec.php [hunk 11]
```diff
@@ -21,11 +21,34 @@
...
+	// Mock test
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: remove this comment, need clear part when mock actually happen in what line

File: spec/App/MarkdownSpec.php [hunk 12]
```diff
@@ -21,11 +21,34 @@
...
+        $this->beConstructedWith($writer);
+
+        // Describe method App\Writer::writeText() should called before example
+    	$writer->writeText('Hi, there')->shouldBeCalled();
+
+    	// when
+        $this->outputHtml('Hi, there', $writer);
+    }
+
+    // Stub test
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: remove this comment, need clear part when stub actually happen in what line

**Comment by samsonasik** [comment]
please update PR title with something like: "more MarkdownSpec specs and other spec sample(movie)" and its PR description

File: spec/App/MovieSpec.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace spec\App;
+
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
```
Comment (samsonasik) [inline spec/App/MovieSpec.php]: this `use Prophecy\Argument;` can be removed if not needed.

File: spec/App/MarkdownSpec.php [hunk 14]
```diff
@@ -21,11 +21,34 @@
 
 use PhpSpec\ObjectBehavior;
 use Prophecy\Argument;
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: this `use Prophecy\Argument;` can be removed if not needed.

**Comment by samsonasik** [comment]
any update?

**Comment by ghamoron** [comment]
There is some update, travis become green :)

File: src/App/MoviesInstance.php [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace App;
+
+use App\InvalidArgumentException;
+
+class MoviesInstance
```
Comment (samsonasik) [inline src/App/MoviesInstance.php]: what is it for? Anyway, I disagree for having "Instance" suffix for class name, as "instance" means an object, that created by class instantiation

File: spec/App/MarkdownSpec.php [hunk 16]
```diff
@@ -21,11 +21,32 @@
...
+        $this->beConstructedWith('createForWriting', $reader);
+
+        // Stub is a process in test that record info about function call.
+    	$reader->getMarkdown()->willReturn("<p>Hi, there!</p>");
+
+        // Describe method toHtmlFromReader() must exact <p>Hi, there!</p> or will be return failed
+    	$this->toHtmlFromReader($reader)->shouldReturn("<p>Hi, there!</p>");
     }
-}
+}
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: add new empty line for end of file

File: src/App/Movie.php [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $desc = "Inexplicably popular children's film";
+
+        return $desc;
+    }
+
+    public function setRating($argument1)
+    {
+        if($argument1 < 0) {
+        	throw new InvalidArgumentException("Error Processing Request");
```
Comment (samsonasik) [inline src/App/Movie.php]: use 4 space

File: src/App/Movie.php [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    public function getDescription()
+    {
+        $desc = "Inexplicably popular children's film";
+
+        return $desc;
+    }
+
+    public function setRating($argument1)
+    {
+        if($argument1 < 0) {
```
Comment (samsonasik) [inline src/App/Movie.php]: space after if

**Comment by samsonasik** [comment]
please use php-cs-fixer with PSR2 level to fix the coding style, thanks.

File: spec/App/MarkdownSpec.php [hunk 19]
```diff
@@ -21,11 +21,32 @@
...
+    	$writer->writeText('Hi, there')->shouldBeCalled();
+
+    	// when
+        $this->outputHtml('Hi, there', $writer);
+    }
+
+    function it_converts_text_from_an_external_source(Reader $reader)
+    {
+        // call constructor through method createForWriting
+        $this->beConstructedWith('createForWriting', $reader);
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: should use `Writer` instead of `Reader` no ?

File: spec/App/MarkdownSpec.php [hunk 20]
```diff
@@ -21,11 +21,32 @@
...
+        $this->beConstructedWith($writer);
+
+        // Describe method App\Writer::writeText() should called before example
+    	$writer->writeText('Hi, there')->shouldBeCalled();
+
+    	// when
+        $this->outputHtml('Hi, there', $writer);
+    }
+
+    function it_converts_text_from_an_external_source(Reader $reader)
```
Comment (samsonasik) [inline spec/App/MarkdownSpec.php]: need to pass `Writer` too as it needed in `createForWriting` as I commented in https://github.com/sitrunlab/learn-phpspec/pull/2/files#r41588508

File: src/App/Markdown.php [hunk 21]
```diff
@@ -21,4 +21,27 @@
 
 class Markdown
 {
+    public function __construct($argument1)
+    {
+        // TODO: write logic here
+    }
+
+    public static function createForWriting($argument1)
```
Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` type hint

File: src/App/Markdown.php [hunk 22]
```diff
@@ -21,4 +21,27 @@
...
+
+        return $markdown;
+    }
+
+    public function outputHtml($argument1, $argument2)
+    {
+        $argument2->writeText($argument1);
+    }
+
+    public function toHtmlFromReader($argument1)
```
Comment (samsonasik) [inline src/App/Markdown.php]: add `Reader` type hint

File: src/App/Markdown.php [hunk 23]
```diff
@@ -21,4 +21,27 @@
...
+    public static function createForWriting($argument1)
+    {
+        $markdown = new Markdown($argument1);
+
+        // TODO: write logic here
+
+        return $markdown;
+    }
+
+    public function outputHtml($argument1, $argument2)
```
Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` type hint for 2nd parameter

File: src/App/MoviesInstance.php (hunk 15 above)
Comment (ghamoron) [inline src/App/MoviesInstance.php]: This just for sample when using $this->shouldBeAnInstanceOf('App\Movie');
but now I removed it

File: src/App/Markdown.php [hunk 24]
```diff
@@ -21,4 +21,27 @@
 
 class Markdown
 {
+    public function __construct($argument1)
+    {
+        // TODO: write logic here
+    }
+
+    public static function createForWriting(Reader $argument1)
```
Comment (samsonasik) [inline src/App/Markdown.php]: use `Writer` type hint instead

File: src/App/Markdown.php [hunk 25]
```diff
@@ -21,4 +21,27 @@
 
 class Markdown
 {
+    public function __construct($argument1)
```
Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` type hint berdasarkan : `$this->beConstructedWith($writer);` di specnya

Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` type hint based on `$this->beConstructedWith($writer);` in spec

Comment (samsonasik) [inline src/App/Markdown.php]: add `Writer` type hint based on `$this->beConstructedWith($writer);` in spec

**Comment by samsonasik** [comment]
as expected, StyleCi now error, you can fix with php-cs-fixer or phpcs locally

File: src/App/Movie.php [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    public function hasSoundtrack()
+    {
+        return true;
+    }
+
+    public function getDirectors()
+    {
+        $directors = array('Director one', 'Director two', 'Director three', 'Director four');
```
Comment (samsonasik) [inline src/App/Movie.php]: please use short array syntax

File: src/App/Movie.php [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+        
+        return $directors;
+    }
+
+    public function getCast()
+    {
+        $casts = array('Cast one', 'Cast two', 'Cast three', 'Jane Smith');
```
Comment (samsonasik) [inline src/App/Movie.php]: short array syntax `['Cast one', 'Cast two', 'Cast three', 'Jane Smith']`

File: spec/App/MovieSpec.php [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+    {
+    	// It compares return data with operator '=='
+
+    	// describe method App\Movie::getRatingLike() should return 5 or "5", if not it will be failed
+        $this->getRatingLike()->shouldBeLike(5);
+    }
+
+    function it_should_not_allow_negative_ratings()
+    {
+    	// It will be failed if method setRating is negatif number
```
Comment (samsonasik) [inline spec/App/MovieSpec.php]: grammar: /s/negatif/negative

File: src/App/Markdown.php [hunk 29]
```diff
@@ -21,4 +21,27 @@
 
 class Markdown
 {
+    public function __construct(Writer $argument1)
```
Comment (samsonasik) [inline src/App/Markdown.php]: change `$argument1` with appropriate name, like `$writer`, etc. Other methods arguments need to be updated too.

File: spec/App/MovieSpec.php [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+{
+    function it_should_be_a_great_movie()
+    {
+        // It compares return data with operator '==='
+
+        // describe method App\Movie::getRating() should return '5' and type of data is integer, if not it will be failed
+        $this->getRating()->shouldBe(5);
+
+        // It should return "Star Wars"
+        $this->getTitle()->shouldBeEqualTo("Star Wars");
```
Comment (samsonasik) [inline spec/App/MovieSpec.php]: single quote

File: spec/App/MovieSpec.php [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+
+        // It should return integer 233366400
+        $this->getReleaseDate()->shouldReturn(233366400);
+
+        // It should exact "Inexplicably popular children's film"
+        $this->getDescription()->shouldEqual("Inexplicably popular children's film");
```
Comment (samsonasik) [inline spec/App/MovieSpec.php:39]: single quote

File: src/App/Movie.php [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+    public function getRatingLike()
+    {
+        $rating = "5";
+        
+        return $rating;
+    }
+
+    public function getTitle()
+    {
+        $title = "Star Wars";
```
Comment (samsonasik) [inline src/App/Movie.php]: here too, single quote

**Pull request merged**